lib/puppet/resource/type.rb

```diff
@@ -196,7 +196,7 @@ def instantiate_resource(scope, resource)
 
   def name
     if type == :node && name_is_regex?
-      "__node_regexp__#{@name.source.downcase.gsub(/[^-\w:.]/, '').sub(/^\.+/, '')}"
+      "__node_regexp__#{@name.source.downcase.gsub(/\(\?[^)]*\)/, '').gsub(/[^-\w:.]/, '').sub(/^\.+/, '')}"
```

Comment:
As you mentioned in the commit message, this results in:

```
irb(main):014:0> "__node_regexp__#{/(?<!a)sync/.source.downcase.gsub(/\(\?[^)]*\)/, '').gsub(/[^-\w:.]/, '').sub(/^\.+/, '')}"
=> "__node_regexp__sync"
```

So it no longer conflicts with the mapping for `node /async/`. But it would now conflict with `node /sync/`

```
irb(main):014:0> "__node_regexp__#{/sync/.source.downcase.gsub(/\(\?[^)]*\)/, '').gsub(/[^-\w:.]/, '').sub(/^\.+/, '')}"
=> "__node_regexp__sync"
```

I think the ideal solution would be to map each invalid character with something that is valid, such as replacing `?` with the string `quote`, but it would need to be marked in such a way as to not conflict with `node 'quote'`